talk/expert/coroutines.tex

```diff
@@ -0,0 +1,551 @@
+      \node[call]   at (1,-7) (body) {coroutine body};
+      \draw (co_await)      edge (await_ready)
+            (await_ready)   edge node[above] {false}  (suspend)
+            (await_ready.west) -- node[above] {true} +(-0.1,0) |- (body);
+      \draw (suspend)       -- +(0,-.5)  -|   (await_suspend);
+      \draw (await_suspend) -- node[above,pos=.1] {false} +(-.95,0) |- (resumed);
+      \draw (await_suspend) edge node[text width=1cm,left,pos=.3] {coroutine handle} (handleresume)
+            (handleresume)  edge (corofin)
+            (corofin)       edge (resumed)
+            (await_suspend) |- node[right,pos=0.1] {\texttt{true}}  (caller)
+            (caller)        edge (callresume)
+            (callresume)    --   +(0,-.5) -| (resumed)
+            (await_suspend) -|  node[right,pos=0.6] {\texttt{void}}  (caller)
+            (resumed)       edge (await_resume)
+            (await_resume)       edge (body);
+      \node[call,minimum width=1cm] at (.7, 0) {user code};
+      \node[comp,minimum width=1cm] at (1, 0) {compiler};
+      \node[box, minimum width=1cm] at (1.3, 0) {awaitable};
+    \end{tikzpicture}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{\texttt{awaiter} example}
+  \begin{block}{Switch to new thread - \href{https://godbolt.org/z/qhvnnf5W3}{\color{blue!50}godbolt}, from \href{https://en.cppreference.com/w/cpp/language/coroutines}{\color{blue!50}cppreference}}
+    \scriptsize
+    \begin{cppcode*}{gobble=2}
+      auto switch_to_new_thread(std::jthread& out) {
+        struct awaitable {
+          std::jthread* p_out;
+          bool await_ready() { return false; }
+          void await_suspend(std::coroutine_handle<> h) {
+            std::jthread& out = *p_out;
+            out = std::jthread([h] { h.resume(); });
+            std::cout << "New thread ID: " << out.get_id() << '\n';
+          }
+          void await_resume() {}
+        };
+        return awaitable{&out};
+      }
+      Task<int> resuming_on_new_thread(std::jthread& out) {
+        std::cout << "Started on " << std::this_thread::get_id() << '\n';
+        co_await switch_to_new_thread(out);
+        std::cout << "Resumed on " << std::this_thread::get_id() << std::endl;
+      }
+      int main() {
+        std::jthread out;
+        resuming_on_new_thread(out);
+      }
+     \end{cppcode*}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Summary on coroutine}
+  \begin{exampleblock}{Coroutines are usable in \cpp20}
+    \begin{itemize}
+    \item allow to have nice generators
+    \item also very useful for async/IO code
+      \begin{itemize}
+      \item another big use case for this we did not touch
+      \end{itemize}
+    \end{itemize}
+  \end{exampleblock}
+  \begin{alertblock}{But library support is poor for now}
+    \begin{itemize}
+    \item standard Task/Generator classes are not provided
+    \item leading to lots of boiler plate code for the end user
```

Comment:
In exchange for removing some mentioning of "boiler plate" before, you may emphasis it here:
```suggestion
    \item leading to \emph{lots of boiler plate code} for the end user
```